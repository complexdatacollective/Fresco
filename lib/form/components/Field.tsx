'use client';

import { motion } from 'motion/react';
import { type ElementType } from 'react';
import { cva } from '~/utils/cva';
import { useField, type UseFieldKeys } from '../hooks/useField';
import {
  type BaseFieldProps,
  type FieldValidation,
  type FieldValue,
} from '../types';
import { getInputState } from '../utils/getInputState';
import FieldErrors from './FieldErrors';
import Hint from './Hint';
import { Label } from './Label';

export const containerVariants = cva({
  base: 'grid gap-3 not-first:mt-6',
  variants: {
    state: {
      valid: 'border-success',
      warning: 'border-warning',
      invalid: 'border-destructive',
      disabled: 'opacity-50 cursor-not-allowed',
    },
  },
});

// Animation variants for field mount/unmount and layout changes
export const fieldAnimationVariants = {
  initial: {
    opacity: 0,
  },
  animate: {
    opacity: 1,
  },
  exit: {
    opacity: 0,
  },
};

/**
 * Wrapper that connects a field to the form context, and handles validation
 * and state.
 *
 * additionalFieldProps should be typed to match the specific field component
 * being used, allowing for additional props to be passed through.
 */
export default function Field<
  TComponent extends React.ElementType,
  TComponentProps = React.ComponentPropsWithoutRef<TComponent>,
>({
  name,
  label,
  hint,
  initialValue,
  validation,
  Component,
  ...additionalFieldProps
}: {
  name: string;
  initialValue?: FieldValue;
  validation?: FieldValidation;
  Component: TComponent;
} & BaseFieldProps &
  Omit<TComponentProps, UseFieldKeys>) {
  // Determine if field is required from validation
  const isRequired = validation
    ? typeof validation === 'function'
      ? false // Can't determine from function, default to false
      : !validation.isOptional()
    : false;

  const { id, meta, fieldProps, containerProps } = useField({
    name,
    initialValue,
    validation,
    required: isRequired,
  });

  const FieldComponent = Component as ElementType;

  const inputVariantState = getInputState(meta);

  return (
    <motion.div
      key={id}
      className={containerVariants({ state: inputVariantState })}
      variants={fieldAnimationVariants}
      layoutId={name}
      initial="initial"
      animate="animate"
      exit="exit"
      layout="position"
      {...containerProps}
    >
      <div>
        <Label id={`${id}-label`} htmlFor={id}>
          {label}
        </Label>
        {hint && (
          <Hint id={`${id}-hint`} validation={validation}>
            {hint}
          </Hint>
        )}
      </div>
      <FieldComponent
        name={name}
        id={id}
        {...fieldProps}
        {...additionalFieldProps}
      />
      <FieldErrors
        id={`${id}-error`}
        errors={meta.errors}
        show={meta.shouldShowError}
      />
    </motion.div>
  );
}
