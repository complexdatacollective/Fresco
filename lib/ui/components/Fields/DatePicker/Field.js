import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Icon from '../../Icon';
import DatePicker from './DatePicker';
import MarkdownLabel from '../MarkdownLabel';

class DatePickerField extends Component {
  constructor(props) {
    super(props);

    this.ref = React.createRef();
  }

  render() {
    const {
      input,
      meta: { error, invalid, touched },
      label,
      placeholder,
      parameters,
      fieldLabel,
      className,
      hidden,
    } = this.props;

    const formFieldClasses = cx(
      className,
      'form-field-date-picker',
      { 'form-field-date-picker--has-error': invalid && touched },
    );

    const anyLabel = fieldLabel || label;
    return (
      <div className="form-field-container" hidden={hidden} ref={this.ref}>
        {anyLabel
          && <MarkdownLabel label={anyLabel} />}
        <div className={formFieldClasses} name={input.name}>
          <DatePicker
            parameters={parameters}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...input}
            onChange={input.onBlur}
            parentRef={this.ref}
            placeholder={placeholder}
          />
          {invalid && touched
            && (
              <div className="form-field-date-picker__error">
                <div className="form-field-date-picker__error-message">
                  <Icon name="warning" />
                  {error}
                </div>
              </div>
            )}
        </div>

      </div>

    );
  }
}

DatePickerField.propTypes = {
  parameters: PropTypes.object,
  input: PropTypes.object.isRequired,
  meta: PropTypes.object,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  fieldLabel: PropTypes.string,
  className: PropTypes.string,
  hidden: PropTypes.bool,
};

DatePickerField.defaultProps = {
  parameters: {},
  meta: {},
  label: null,
  placeholder: null,
  fieldLabel: null,
  className: null,
  hidden: null,
};

export default DatePickerField;
