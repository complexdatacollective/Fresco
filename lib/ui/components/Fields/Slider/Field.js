import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Icon from '../../Icon';
import Slider from './Slider';
import MarkdownLabel from '../MarkdownLabel';

const getSliderType = (variableType) => {
  switch (variableType) {
    case 'ordinal':
      return 'LIKERT';
    case 'scalar':
      return 'VAS';
    default:
      return null;
  }
};

const hasValue = (value) => value !== '';

/**
 * Empty string value should be treated as `null`
 * because redux-forms turns `null` values (e.g.
 * unset values) into empty strings when
 * building the input object...
 */
const getValue = (value) => {
  if (!hasValue(value)) { return null; }
  return value;
};

const SliderField = (props) => {
  const {
    input,
    meta: { error, invalid, touched },
    label,
    parameters,
    options,
    fieldLabel,
    className,
    hidden,
    type,
  } = props;

  const formFieldClasses = cx(
    className,
    'form-field-slider',
    { 'form-field-slider--has-error': invalid && touched },
  );

  const anyLabel = fieldLabel || label;
  const sliderType = getSliderType(type);

  return (
    <div className="form-field-container" hidden={hidden}>
      { anyLabel
        && <MarkdownLabel label={anyLabel} />}
      <div className={formFieldClasses} name={input.name}>
        <Slider
          options={options}
          parameters={parameters}
          type={sliderType}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...input}
          value={getValue(input.value)}
        />
        {invalid && touched && (
          <div className="form-field-slider__error">
            <Icon name="warning" />
            {error}
          </div>
        )}
      </div>
    </div>
  );
};

SliderField.propTypes = {
  label: PropTypes.node,
  className: PropTypes.string,
  hidden: PropTypes.bool,
  input: PropTypes.object.isRequired,
  meta: PropTypes.object,
  parameters: PropTypes.object,
  options: PropTypes.array,
  fieldLabel: PropTypes.string,
  type: PropTypes.string.isRequired,
};

SliderField.defaultProps = {
  className: '',
  label: null,
  hidden: false,
  meta: {},
  fieldLabel: null,
  options: null,
  parameters: null,
};

export default SliderField;
