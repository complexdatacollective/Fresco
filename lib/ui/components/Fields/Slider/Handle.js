import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import MarkdownLabel from '../MarkdownLabel';

class Handle extends Component {
  constructor(props) {
    super(props);

    this.state = {
      mouseOver: false,
    };
  }

  handleMouseOver = () => {
    this.setState({ mouseOver: true });
  }

  handleMouseLeave = () => {
    this.setState({ mouseOver: false });
  }

  render() {
    const {
      domain: [min, max],
      handle: { id, value, percent },
      isActive,
      isDisabled,
      showTooltips,
      getHandleProps,
      getLabelForValue,
    } = this.props;
    const { mouseOver } = this.state;

    const showTooltip = showTooltips && (mouseOver || isActive) && !isDisabled;
    const handleProps = getHandleProps(
      id,
      {
        onMouseEnter: this.handleMouseEnter,
        onMouseLeave: this.handleMouseLeave,
      },
    );

    const markerClasses = cx(
      'form-field-slider__marker',
      { 'form-field-slider__marker--is-active': isActive },
      { 'form-field-slider__marker--is-disabled': isDisabled },
    );

    const tooltipClasses = cx(
      'form-field-slider__tooltip',
      { 'form-field-slider__tooltip--is-active': showTooltip },
    );

    const label = getLabelForValue(value);

    return (
      <>
        { showTooltips
          && (
          <div
            className={tooltipClasses}
            style={{ left: `${percent}%` }}
          >
            <MarkdownLabel inline label={label} className="form-field-slider__tooltip-label" />
          </div>
          )}
        <div
          className="form-field-slider__handle"
          style={{ left: `${percent}%` }}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...handleProps}
        />
        <div
          role="slider"
          aria-label="Slider"
          aria-valuemin={min}
          aria-valuemax={max}
          aria-valuenow={value}
          className={markerClasses}
          style={{ left: `${percent}%` }}
        />
      </>
    );
  }
}

Handle.propTypes = {
  domain: PropTypes.array.isRequired,
  handle: PropTypes.shape({
    id: PropTypes.any.isRequired,
    value: PropTypes.number.isRequired,
    percent: PropTypes.number.isRequired,
  }).isRequired,
  isActive: PropTypes.bool,
  isDisabled: PropTypes.bool,
  showTooltips: PropTypes.bool,
  getHandleProps: PropTypes.func.isRequired,
  getLabelForValue: PropTypes.func.isRequired,
};

Handle.defaultProps = {
  isActive: false,
  isDisabled: false,
  showTooltips: false,
};

export default Handle;
