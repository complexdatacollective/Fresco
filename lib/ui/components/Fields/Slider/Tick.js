import React from 'react';
import PropTypes from 'prop-types';
import MarkdownLabel from '../MarkdownLabel';

const Tick = ({ tick, getLabelForValue }) => {
  const { value, percent } = tick;
  const label = getLabelForValue(value);

  return (
    <div
      className="form-field-slider__tick"
      style={{
        position: 'absolute',
        left: `${percent}%`,
      }}
    >
      {label && <MarkdownLabel inline label={label} className="form-field-slider__tick-label" />}
    </div>
  );
};

Tick.propTypes = {
  tick: PropTypes.any.isRequired,
  getLabelForValue: PropTypes.func,
};

Tick.defaultProps = {
  getLabelForValue: () => null,
};

export default Tick;
