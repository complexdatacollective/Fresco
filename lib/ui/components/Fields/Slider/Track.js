import React from 'react';
import PropTypes from 'prop-types';

const Track = ({ source, target, getTrackProps }) => (
  <div
    className="form-field-slider__track"
    style={{
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    // eslint-disable-next-line react/jsx-props-no-spreading
    {...getTrackProps()}
  >
    <div className="form-field-slider__track-line" />
  </div>
);

Track.propTypes = {
  source: PropTypes.object.isRequired,
  target: PropTypes.object.isRequired,
  getTrackProps: PropTypes.func.isRequired,
};

export default Track;
