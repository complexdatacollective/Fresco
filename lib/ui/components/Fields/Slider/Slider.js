import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { round, get, isNil } from 'lodash';
import cx from 'classnames';
import {
  Slider, Handles, Tracks, Ticks,
} from 'react-compound-slider';
import Handle from './Handle';
import Track from './Track';
import Tick from './Tick';

class SliderInput extends Component {
  getSliderProps = () => {
    const { options, value } = this.props;

    const domain = this.isLikert()
      ? [0, options.length - 1]
      : [0, 1];

    const step = this.isLikert() ? 1 : 0.0005;

    const values = this.isLikert()
      ? [options.findIndex((option) => option.value === value)]
      : [value];

    return {
      domain,
      step,
      values,
    };
  };

  getTickCount = () => {
    const { type, options } = this.props;

    switch (type) {
      case 'LIKERT':
        return options.length - 1;
      case 'VAS':
        return 1;
      default:
        return null;
    }
  };

  getLabelForValue = (value) => {
    const { options, parameters } = this.props;
    if (this.isLikert()) { return get(options, [value, 'label']); }
    if (this.isVisualAnalogScale()) {
      const index = value === 0 ? 'minLabel' : 'maxLabel';
      return get(parameters, index);
    }
    return round(value * 100);
  }

  normalizeValue = (value) => {
    const {
      options,
    } = this.props;

    if (this.isLikert()) {
      return options[value].value;
    }
    return round(value, 3);
  }

  /**
   * The onChange property is called on initialization, so
   * we are using handleSlideEnd() to capture changes.
   */
  handleSlideEnd = (value) => {
    const { onBlur } = this.props;
    const normalizedValue = this.normalizeValue(value);
    // Use input.onBlur rather than input.onChange so that we can set 'touched'
    onBlur(normalizedValue);
  }

  isLikert = () => {
    const { type } = this.props;
    return type === 'LIKERT';
  }

  isVisualAnalogScale = () => {
    const { type } = this.props;
    return type === 'VAS';
  };

  render() {
    const {
      value,
    } = this.props;

    const sliderProps = this.getSliderProps();
    const tickCount = this.getTickCount();
    const showTooltips = !this.isVisualAnalogScale();
    const isNotSet = isNil(value);

    const className = cx(
      'form-field-slider__slider',
      { 'form-field-slider__slider--likert': this.isLikert() },
      { 'form-field-slider__slider--vas': this.isVisualAnalogScale() },
      { 'form-field-slider__slider--not-set': isNotSet },
    );

    return (
      <div className="form-field">
        <Slider
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...sliderProps}
          className={className}
          onSlideEnd={this.handleSlideEnd}
        >
          <Handles>
            {({ handles, activeHandleID, getHandleProps }) => (
              <div className="form-field-slider__handles">
                {handles.map((handle) => (
                  <Handle
                    key={handle.id}
                    handle={handle}
                    getLabelForValue={this.getLabelForValue}
                    domain={sliderProps.domain}
                    isActive={handle.id === activeHandleID}
                    getHandleProps={getHandleProps}
                    showTooltips={showTooltips}
                  />
                ))}
              </div>
            )}
          </Handles>
          <Tracks>
            {({ tracks, getTrackProps }) => (
              <div className="form-field-slider__tracks">
                {tracks.map(({ id, source, target }) => (
                  <Track
                    key={id}
                    source={source}
                    target={target}
                    getTrackProps={getTrackProps}
                  />
                ))}
              </div>
            )}
          </Tracks>
          { tickCount
            && (
            <Ticks count={tickCount}>
              {({ ticks }) => (
                <div className="form-field-slider__ticks">
                  {ticks.map((tick, index) => (
                    <Tick
                      tick={tick}
                      key={`${this.getLabelForValue}_${index}`}
                      getLabelForValue={this.getLabelForValue}
                    />
                  ))}
                </div>
              )}
            </Ticks>
            )}
        </Slider>
      </div>
    );
  }
}

SliderInput.propTypes = {
  options: PropTypes.array,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  type: PropTypes.string.isRequired,
  onBlur: PropTypes.func.isRequired,
  parameters: PropTypes.shape({
    minLabel: PropTypes.string,
    maxLabel: PropTypes.string,
  }),
};

SliderInput.defaultProps = {
  options: [],
  value: null,
  parameters: {},
};

export default SliderInput;
