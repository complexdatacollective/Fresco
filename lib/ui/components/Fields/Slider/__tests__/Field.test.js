/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import Field from '../Field';
import Slider from '../Slider';
import Tick from '../Tick';
import Handle from '../Handle';

const getMockProps = (props) => ({
  input: { onChange: () => {} },
  meta: { error: null, invalid: null, touched: null },
  ...props,
});

describe('Slider/Field', () => {
  it('Correctly renders LIKERT slider (ordinal variable)', () => {
    const mockProps = getMockProps({
      type: 'ordinal',
      options: [
        { value: 'FOO', label: 'foo' },
        { value: 'BAR', label: 'bar' },
        { value: 'BAZZ', label: 'bazz' },
        { value: 'FIZZ', label: 'fizz' },
      ],
    });
    // eslint-disable-next-line react/jsx-props-no-spreading
    const subject = mount((<Field {...mockProps} />));
    expect(subject.html()).toMatchSnapshot();

    const slider = subject.find(Slider);
    expect(slider.prop('type')).toEqual('LIKERT');

    const ticks = subject.find(Tick);
    expect(ticks).toHaveLength(4);

    const handle = subject.find(Handle);
    handle.find('.form-field-slider__handle').simulate('mousedown');
    const tooltip = handle.find('.form-field-slider__tooltip');
    expect(tooltip.html()).toMatchSnapshot();
  });

  it('Correctly renders VISUAL ANALOG SCALE slider (scalar variable)', () => {
    const mockProps = getMockProps({
      type: 'scalar',
      options: {
        minLabel: 'Fizz',
        maxLabel: 'Bazz',
      },
    });
    // eslint-disable-next-line react/jsx-props-no-spreading
    const subject = mount((<Field {...mockProps} />));
    expect(subject.html()).toMatchSnapshot();

    const slider = subject.find(Slider);
    expect(slider.prop('type')).toEqual('VAS');

    const ticks = subject.find(Tick);
    expect(ticks).toHaveLength(2);

    const handle = subject.find(Handle);
    handle.find('.form-field-slider__handle').simulate('mousedown');

    expect(handle.exists('.form-field-slider__tooltip')).toBe(false);
  });

  it('Correctly renders slider (number variable)', () => {
    const mockProps = getMockProps({
      type: 'number',
    });
    // eslint-disable-next-line react/jsx-props-no-spreading
    const subject = mount((<Field {...mockProps} />));
    expect(subject.html()).toMatchSnapshot();

    const slider = subject.find(Slider);
    expect(slider.prop('type')).toEqual(null);

    const ticks = subject.find(Tick);
    expect(ticks).toHaveLength(0);

    const handle = subject.find(Handle);
    handle.find('.form-field-slider__handle').simulate('mousedown');
    const tooltip = handle.find('.form-field-slider__tooltip');
    expect(tooltip.html()).toMatchSnapshot();
  });
});
