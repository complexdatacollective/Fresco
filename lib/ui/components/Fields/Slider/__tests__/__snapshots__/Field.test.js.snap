// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider/Field Correctly renders LIKERT slider (ordinal variable) 1`] = `"<div class=\\"form-field-container\\"><div class=\\"form-field-slider\\"><div class=\\"form-field-slider__slider form-field-slider__slider--likert form-field-slider__slider--not-set\\"><div class=\\"form-field-slider__handles\\"><div class=\\"form-field-slider__tooltip\\" style=\\"left: 0%;\\"><div class=\\"form-field-slider__tooltip-label\\">foo</div></div><div class=\\"form-field-slider__handle\\" style=\\"left: 0%;\\"></div><div role=\\"slider\\" aria-valuemin=\\"0\\" aria-valuemax=\\"3\\" aria-valuenow=\\"0\\" class=\\"form-field-slider__marker\\" style=\\"left: 0%;\\"></div></div><div class=\\"form-field-slider__tracks\\"><div class=\\"form-field-slider__track\\" style=\\"left: 0%; width: 0%;\\"><div class=\\"form-field-slider__track-line\\"></div></div><div class=\\"form-field-slider__track\\" style=\\"left: 0%; width: 100%;\\"><div class=\\"form-field-slider__track-line\\"></div></div></div><div class=\\"form-field-slider__ticks\\"><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 0%;\\"><div class=\\"form-field-slider__tick-label\\">foo</div></div><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 33.33333333333333%;\\"><div class=\\"form-field-slider__tick-label\\">bar</div></div><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 66.66666666666666%;\\"><div class=\\"form-field-slider__tick-label\\">bazz</div></div><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 100%;\\"><div class=\\"form-field-slider__tick-label\\">fizz</div></div></div></div><div class=\\"form-field-slider__error\\"><div class=\\"form-field-slider__error-message\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 180 160\\" class=\\"icon\\" name=\\"warning\\"><title>Warning</title><circle class=\\"cls-1\\" cx=\\"90\\" cy=\\"17.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-1\\" cx=\\"17.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-2\\" cx=\\"162.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><path class=\\"cls-1\\" d=\\"M1.84 134.68L74.19 10h31.56l72.41 124.68H1.84z\\"></path><path class=\\"cls-2\\" d=\\"M78 134.68l63.36-63.36 36.8 63.36H78z\\"></path><path class=\\"cls-1\\" d=\\"M17.5 126.68h145V160h-145z\\"></path><circle class=\\"cls-3\\" cx=\\"84.29\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-3\\" cx=\\"95.71\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"79.93\\" cy=\\"48.69\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"100.07\\" cy=\\"48.69\\" r=\\"3\\"></circle><path class=\\"cls-3\\" d=\\"M84.29 96.31h11.42v6H84.29z\\"></path><path class=\\"cls-4\\" d=\\"M79.93 45.69h20.15v6H79.93z\\"></path><path class=\\"cls-4\\" d=\\"M76.93 48.69l4.36 50.62h17.42l4.36-50.62H76.93z\\"></path><path class=\\"cls-3\\" d=\\"M80.03 84.75l1.26 14.56h17.42l3.13-36.37-21.81 21.81z\\"></path><path class=\\"cls-2\\" d=\\"M162.5 126.68H86L52.64 160H162.5v-33.32z\\"></path><circle class=\\"cls-4\\" cx=\\"90\\" cy=\\"122.68\\" r=\\"12\\"></circle><path class=\\"cls-3\\" d=\\"M81.51 131.17a12 12 0 0 0 17-17z\\"></path></svg></div></div></div></div>"`;

exports[`Slider/Field Correctly renders LIKERT slider (ordinal variable) 2`] = `"<div class=\\"form-field-slider__tooltip form-field-slider__tooltip--is-active\\" style=\\"left: 0%;\\"><div class=\\"form-field-slider__tooltip-label\\">foo</div></div>"`;

exports[`Slider/Field Correctly renders VISUAL ANALOG SCALE slider (scalar variable) 1`] = `"<div class=\\"form-field-container\\"><div class=\\"form-field-slider\\"><div class=\\"form-field-slider__slider form-field-slider__slider--vas form-field-slider__slider--not-set\\"><div class=\\"form-field-slider__handles\\"><div class=\\"form-field-slider__handle\\"></div><div role=\\"slider\\" aria-valuemin=\\"0\\" aria-valuemax=\\"1\\" aria-valuenow=\\"NaN\\" class=\\"form-field-slider__marker\\"></div></div><div class=\\"form-field-slider__tracks\\"><div class=\\"form-field-slider__track\\" style=\\"left: 0%;\\"><div class=\\"form-field-slider__track-line\\"></div></div><div class=\\"form-field-slider__track\\"><div class=\\"form-field-slider__track-line\\"></div></div></div><div class=\\"form-field-slider__ticks\\"><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 0%;\\"><div class=\\"form-field-slider__tick-label\\"></div></div><div class=\\"form-field-slider__tick\\" style=\\"position: absolute; left: 100%;\\"><div class=\\"form-field-slider__tick-label\\"></div></div></div></div><div class=\\"form-field-slider__error\\"><div class=\\"form-field-slider__error-message\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 180 160\\" class=\\"icon\\" name=\\"warning\\"><title>Warning</title><circle class=\\"cls-1\\" cx=\\"90\\" cy=\\"17.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-1\\" cx=\\"17.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-2\\" cx=\\"162.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><path class=\\"cls-1\\" d=\\"M1.84 134.68L74.19 10h31.56l72.41 124.68H1.84z\\"></path><path class=\\"cls-2\\" d=\\"M78 134.68l63.36-63.36 36.8 63.36H78z\\"></path><path class=\\"cls-1\\" d=\\"M17.5 126.68h145V160h-145z\\"></path><circle class=\\"cls-3\\" cx=\\"84.29\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-3\\" cx=\\"95.71\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"79.93\\" cy=\\"48.69\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"100.07\\" cy=\\"48.69\\" r=\\"3\\"></circle><path class=\\"cls-3\\" d=\\"M84.29 96.31h11.42v6H84.29z\\"></path><path class=\\"cls-4\\" d=\\"M79.93 45.69h20.15v6H79.93z\\"></path><path class=\\"cls-4\\" d=\\"M76.93 48.69l4.36 50.62h17.42l4.36-50.62H76.93z\\"></path><path class=\\"cls-3\\" d=\\"M80.03 84.75l1.26 14.56h17.42l3.13-36.37-21.81 21.81z\\"></path><path class=\\"cls-2\\" d=\\"M162.5 126.68H86L52.64 160H162.5v-33.32z\\"></path><circle class=\\"cls-4\\" cx=\\"90\\" cy=\\"122.68\\" r=\\"12\\"></circle><path class=\\"cls-3\\" d=\\"M81.51 131.17a12 12 0 0 0 17-17z\\"></path></svg></div></div></div></div>"`;

exports[`Slider/Field Correctly renders slider (number variable) 1`] = `"<div class=\\"form-field-container\\"><div class=\\"form-field-slider\\"><div class=\\"form-field-slider__slider form-field-slider__slider--not-set\\"><div class=\\"form-field-slider__handles\\"><div class=\\"form-field-slider__tooltip\\"><div class=\\"form-field-slider__tooltip-label\\">NaN</div></div><div class=\\"form-field-slider__handle\\"></div><div role=\\"slider\\" aria-valuemin=\\"0\\" aria-valuemax=\\"1\\" aria-valuenow=\\"NaN\\" class=\\"form-field-slider__marker\\"></div></div><div class=\\"form-field-slider__tracks\\"><div class=\\"form-field-slider__track\\" style=\\"left: 0%;\\"><div class=\\"form-field-slider__track-line\\"></div></div><div class=\\"form-field-slider__track\\"><div class=\\"form-field-slider__track-line\\"></div></div></div></div><div class=\\"form-field-slider__error\\"><div class=\\"form-field-slider__error-message\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 180 160\\" class=\\"icon\\" name=\\"warning\\"><title>Warning</title><circle class=\\"cls-1\\" cx=\\"90\\" cy=\\"17.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-1\\" cx=\\"17.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><circle class=\\"cls-2\\" cx=\\"162.5\\" cy=\\"142.5\\" r=\\"17.5\\"></circle><path class=\\"cls-1\\" d=\\"M1.84 134.68L74.19 10h31.56l72.41 124.68H1.84z\\"></path><path class=\\"cls-2\\" d=\\"M78 134.68l63.36-63.36 36.8 63.36H78z\\"></path><path class=\\"cls-1\\" d=\\"M17.5 126.68h145V160h-145z\\"></path><circle class=\\"cls-3\\" cx=\\"84.29\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-3\\" cx=\\"95.71\\" cy=\\"99.31\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"79.93\\" cy=\\"48.69\\" r=\\"3\\"></circle><circle class=\\"cls-4\\" cx=\\"100.07\\" cy=\\"48.69\\" r=\\"3\\"></circle><path class=\\"cls-3\\" d=\\"M84.29 96.31h11.42v6H84.29z\\"></path><path class=\\"cls-4\\" d=\\"M79.93 45.69h20.15v6H79.93z\\"></path><path class=\\"cls-4\\" d=\\"M76.93 48.69l4.36 50.62h17.42l4.36-50.62H76.93z\\"></path><path class=\\"cls-3\\" d=\\"M80.03 84.75l1.26 14.56h17.42l3.13-36.37-21.81 21.81z\\"></path><path class=\\"cls-2\\" d=\\"M162.5 126.68H86L52.64 160H162.5v-33.32z\\"></path><circle class=\\"cls-4\\" cx=\\"90\\" cy=\\"122.68\\" r=\\"12\\"></circle><path class=\\"cls-3\\" d=\\"M81.51 131.17a12 12 0 0 0 17-17z\\"></path></svg></div></div></div></div>"`;

exports[`Slider/Field Correctly renders slider (number variable) 2`] = `"<div class=\\"form-field-slider__tooltip form-field-slider__tooltip--is-active\\"><div class=\\"form-field-slider__tooltip-label\\">NaN</div></div>"`;
